File: backend/app/domain/services/recommendation.py
from app.domain.services.base import BaseService
import uuid
from datetime import datetime, timezone
from typing import Any, Optional

from sqlalchemy.orm import Session

from app.core.exceptions import NotFoundError, ValidationError
from app.core.redis import (
    cache_org_recommendations,
    get_cached_org_recommendations,
    invalidate_org_recommendations,
)
from app.domain.entities.portfolio import Transaction
from app.domain.entities.recommendation import (
    Recommendation,
    RecommendationExplanation,
    RecommendationOutcome,
)
from app.infrastructure.repositories.recommendation import RecommendationRepository


def serialize_recommendation(rec: Recommendation, session: Session) -> dict[str, Any]:
    repo = RecommendationRepository(session)
    expl = repo.get_explanation(rec.id)
    out = repo.get_outcome(rec.id)
    quote = repo.get_quote_by_asset(rec.asset_id)
    symbol = quote.symbol if quote else None
    
    return {
        "id": str(rec.id),
        "organization_id": str(rec.organization_id),
        "asset_id": str(rec.asset_id),
        "symbol": symbol,
        "recommendation_state": rec.recommendation_state,
        "confidence_score": float(rec.confidence_score),
        "status": rec.status,
        "version": rec.version,
        "created_at": rec.created_at.isoformat() if rec.created_at else None,
        "updated_at": rec.updated_at.isoformat() if rec.updated_at else None,
        "explanation": {
            "rules_matched": expl.rules_matched if expl else {},
            "reasoning": expl.reasoning if expl else "",
            "confidence_factors": expl.confidence_factors if expl else {}
        } if expl else None,
        "outcome": {
            "status": out.status if out else "active",
            "action_taken_at": out.action_taken_at.isoformat() if out and out.action_taken_at else None,
            "dismiss_reason": out.dismiss_reason if out else None,
            "ledger_transaction_id": str(out.ledger_transaction_id) if out and out.ledger_transaction_id else None,
            "predicted_impact": float(out.predicted_impact) if out and out.predicted_impact is not None else None,
            "realized_impact": float(out.realized_impact) if out and out.realized_impact is not None else None
        } if out else None
    }

class RecommendationService(BaseService):
    def __init__(self, session: Session):
        self.session = session
        self.repo = RecommendationRepository(session)

    def generate_recommendations(self, organization_id: uuid.UUID) -> list[dict[str, Any]]:
        snapshots = self.repo.list_all_snapshots()
        recs_created = []

        for snap in snapshots:
            asset_id = snap.asset_id
            features = self.repo.get_features(asset_id)
            scores = self.repo.get_latest_score(asset_id)
            
            if not features or not scores:
                continue
                
            momentum = float(features.momentum_score) if features.momentum_score is not None else 0.5
            volatility = float(features.volatility_score) if features.volatility_score is not None else 0.3
            sentiment = float(features.sentiment_score) if features.sentiment_score is not None else 0.5
            
            quality = float(scores.quality_score) if scores.quality_score is not None else 0.8
            valuation = float(scores.valuation_score) if scores.valuation_score is not None else 0.7
            
            # Rule Engine (Deterministic)
            rec_state = "HOLD"
            reasoning = "Asset parameters remain within stable bounds. Recommending holding current position."
            rules_matched = {}
            confidence_factors = {}
            confidence_score = 0.5
            
            if valuation >= 0.7 and momentum >= 0.5 and sentiment >= 0.5:
                rec_state = "BUY"
                reasoning = "Asset displays strong underpricing combined with positive momentum and constructive market sentiment."
                rules_matched = {"underpricing_and_momentum": True}
                confidence_factors = {"valuation": 0.4, "momentum": 0.3, "sentiment": 0.3}
                confidence_score = 0.4 * valuation + 0.3 * momentum + 0.3 * sentiment
            elif sentiment < 0.3 and momentum < 0.4:
                rec_state = "AVOID"
                reasoning = "Asset displays weak market sentiment and negative momentum, prompting caution."
                rules_matched = {"weak_sentiment_and_momentum": True}
                confidence_factors = {"sentiment": 0.5, "momentum": 0.5}
                confidence_score = 0.5 * (1.0 - sentiment) + 0.5 * (1.0 - momentum)
            elif valuation < 0.4 and volatility >= 0.6:
                rec_state = "REDUCE"
                reasoning = "Asset is potentially overvalued with elevated volatility, recommending reducing exposure."
                rules_matched = {"overvaluation_and_volatility": True}
                confidence_factors = {"valuation": 0.5, "volatility": 0.5}
                confidence_score = 0.5 * (1.0 - valuation) + 0.5 * volatility
            else:
                rec_state = "HOLD"
                reasoning = "Asset parameters remain within stable bounds. Recommending holding current position."
                rules_matched = {"stable_parameters": True}
                confidence_factors = {"quality": 0.5, "volatility": 0.5}
                confidence_score = 0.5 * quality + 0.5 * (1.0 - volatility)
                
            existing_rec = self.repo.get_active_recommendation(organization_id, asset_id, "v2.0.0")

            rec_id = existing_rec.id if existing_rec else uuid.uuid4()
            
            rec = Recommendation(
                id=rec_id,
                organization_id=organization_id,
                asset_id=asset_id,
                recommendation_state=rec_state,
                confidence_score=confidence_score,
                status="active",
                version="v2.0.0",
                created_at=existing_rec.created_at if existing_rec else datetime.now(timezone.utc),
                updated_at=datetime.now(timezone.utc)
            )
            self.repo.upsert(rec)
            
            expl = RecommendationExplanation(
                recommendation_id=rec_id,
                rules_matched=rules_matched,
                reasoning=reasoning,
                confidence_factors=confidence_factors
            )
            self.repo.upsert_explanation(expl)
            
            existing_outcome = self.repo.get_outcome(rec_id)
            if not existing_outcome:
                out = RecommendationOutcome(
                    recommendation_id=rec_id,
                    status="active",
                    action_taken_at=datetime.now(timezone.utc)
                )
                self.repo.upsert_outcome(out)
                
            recs_created.append(rec)
            
        self.session.commit()
        invalidate_org_recommendations(str(organization_id))
        return [serialize_recommendation(r, self.session) for r in recs_created]

    def get_recommendations(self, organization_id: uuid.UUID, status: str | None = None) -> list[dict[str, Any]]:
        cached = get_cached_org_recommendations(str(organization_id))
        if cached is not None:
            if status:
                return [r for r in cached if r.get("status") == status]
            return cached
            
        recs = self.repo.get_by_org(organization_id)
        serialized = [serialize_recommendation(r, self.session) for r in recs]
        cache_org_recommendations(str(organization_id), serialized)
        
        if status:
            return [r for r in serialized if r.get("status") == status]
        return serialized

    def apply_recommendation(self, recommendation_id: uuid.UUID, portfolio_id: uuid.UUID | None = None, actor_id: Optional[uuid.UUID] = None) -> dict[str, Any]:
        rec = self.repo.get(recommendation_id)
        if not rec:
            raise NotFoundError("Recommendation not found")
            
        if rec.status != "active":
            raise ValidationError(f"Recommendation is already {rec.status}")
            
        if not portfolio_id:
            portfolio = self.repo.get_portfolio_by_org(rec.organization_id)
            if not portfolio:
                raise ValidationError("No portfolios found for this organization to apply recommendation")
            portfolio_id = portfolio.id
        else:
            portfolio = self.repo.get_portfolio(portfolio_id, rec.organization_id)
            if not portfolio:
                raise ValidationError("Invalid portfolio for this organization")

        quote = self.repo.get_quote_by_asset(rec.asset_id)
        symbol = quote.symbol if quote else "UNKNOWN"
        price = float(quote.price) if quote else 0.0

        txn = Transaction(
            id=uuid.uuid4(),
            portfolio_id=portfolio_id,
            symbol=symbol,
            asset_id=rec.asset_id,
            transaction_type="BUY" if rec.recommendation_state == "BUY" else "SELL" if rec.recommendation_state in ("REDUCE", "AVOID") else "HOLD",
            quantity=1.0 if rec.recommendation_state in ("BUY", "REDUCE", "AVOID") else 0.0,
            price=price,
            transaction_date=datetime.now(timezone.utc),
            fees=0.0,
            taxes=0.0,
            notes=f"Applied recommendation {rec.id} ({rec.recommendation_state})",
            broker="aureon",
            kind="trade"
        )
        self.repo.add_transaction(txn)
        
        rec.status = "applied"
        rec.updated_at = datetime.now(timezone.utc)
        self.repo.upsert(rec)
        
        out = self.repo.get_outcome(recommendation_id)
        if not out:
            out = RecommendationOutcome(
                recommendation_id=recommendation_id,
                status="applied",
                action_taken_at=datetime.now(timezone.utc)
            )
        out.status = "applied"
        out.action_taken_at = datetime.now(timezone.utc)
        out.ledger_transaction_id = txn.id
        out.predicted_impact = 0.05 if rec.recommendation_state == "BUY" else -0.05 if rec.recommendation_state in ("REDUCE", "AVOID") else 0.0
        self.repo.upsert_outcome(out)
        
        from app.domain.services.audit import log_audit_action
        log_audit_action(
            self.session,
            action="recommendation_apply",
            entity_type="recommendation",
            entity_id=str(recommendation_id),
            actor_id=actor_id,
            details={"recommendation_state": rec.recommendation_state, "portfolio_id": str(portfolio_id), "organization_id": str(rec.organization_id)}
        )
        
        self.session.commit()
        invalidate_org_recommendations(str(rec.organization_id))
        
        try:
            from app.workers.evaluation.scoring import (
                update_financial_intelligence_pipeline,
            )
            update_financial_intelligence_pipeline(self.session, rec.organization_id)
        except Exception:
            pass
        
        return serialize_recommendation(rec, self.session)

    def dismiss_recommendation(self, recommendation_id: uuid.UUID, reason: str | None = None, actor_id: Optional[uuid.UUID] = None) -> dict[str, Any]:
        rec = self.repo.get(recommendation_id)
        if not rec:
            raise NotFoundError("Recommendation not found")
            
        if rec.status != "active":
            raise ValidationError(f"Recommendation is already {rec.status}")
            
        rec.status = "dismissed"
        rec.updated_at = datetime.now(timezone.utc)
        self.repo.upsert(rec)
        
        out = self.repo.get_outcome(recommendation_id)
        if not out:
            out = RecommendationOutcome(
                recommendation_id=recommendation_id,
                status="dismissed",
                action_taken_at=datetime.now(timezone.utc)
            )
        out.status = "dismissed"
        out.action_taken_at = datetime.now(timezone.utc)
        out.dismiss_reason = reason
        self.repo.upsert_outcome(out)
        
        from app.domain.services.audit import log_audit_action
        log_audit_action(
            self.session,
            action="recommendation_dismiss",
            entity_type="recommendation",
            entity_id=str(recommendation_id),
            actor_id=actor_id,
            details={"reason": reason, "organization_id": str(rec.organization_id)}
        )
        
        self.session.commit()
        invalidate_org_recommendations(str(rec.organization_id))
        
        try:
            from app.workers.evaluation.scoring import (
                update_financial_intelligence_pipeline,
            )
            update_financial_intelligence_pipeline(self.session, rec.organization_id)
        except Exception:
            pass
        
        return serialize_recommendation(rec, self.session)

    def undo_recommendation(self, recommendation_id: uuid.UUID, actor_id: Optional[uuid.UUID] = None) -> dict[str, Any]:
        rec = self.repo.get(recommendation_id)
        if not rec:
            raise NotFoundError("Recommendation not found")
            
        if rec.status == "active":
            raise ValidationError("Recommendation is already active")
            
        out = self.repo.get_outcome(recommendation_id)
        if out and out.ledger_transaction_id:
            txn = self.repo.get_transaction(out.ledger_transaction_id)
            if txn:
                self.repo.delete_transaction(txn)
                
        rec.status = "active"
        rec.updated_at = datetime.now(timezone.utc)
        self.repo.upsert(rec)
        
        if out:
            out.status = "active"
            out.action_taken_at = datetime.now(timezone.utc)
            out.ledger_transaction_id = None
            out.dismiss_reason = None
            self.repo.upsert_outcome(out)
            
        from app.domain.services.audit import log_audit_action
        log_audit_action(
            self.session,
            action="recommendation_undo",
            entity_type="recommendation",
            entity_id=str(recommendation_id),
            actor_id=actor_id,
            details={"previous_status": rec.status, "organization_id": str(rec.organization_id)}
        )
            
        self.session.commit()
        invalidate_org_recommendations(str(rec.organization_id))
        
        try:
            from app.workers.evaluation.scoring import (
                update_financial_intelligence_pipeline,
            )
            update_financial_intelligence_pipeline(self.session, rec.organization_id)
        except Exception:
            pass
            
        return serialize_recommendation(rec, self.session)

File: backend/src/lib/ai/recommendation.ts
import { v4 as uuidv4 } from "uuid";
import { prisma } from "../../prisma";
import type { Prisma, recommendations as Recommendation } from "../../generated/prisma";
import { NotFoundError, ValidationError } from "../errors";
import { logAuditAction } from "../audit";
import { invalidateOrgRecommendations } from "../portfolioCache";
import { updateFinancialIntelligencePipeline } from "./intelligence";
import { logger } from "../logger";

// Port of the parts of app/modules/ai/services/recommendation.py needed so
// far: the deterministic rule engine (_score_and_materialize), the
// manual-batch trigger (generate_recommendations), and apply/dismiss/undo.
// apply/dismiss/undo below call update_financial_intelligence_pipeline
// (Task 8, migration plan — see lib/ai/intelligence.ts) wrapped in
// try/catch, exactly matching Python's try/except-wrap-and-continue so a
// pipeline failure never blocks the primary action.

/** Logs and swallows a pipeline failure, matching Python's
 * `except Exception as e: logger.warning(...)` at all 4 call sites. */
async function refreshIntelligencePipeline(operation: string, recommendationId: string, extra?: Record<string, string>): Promise<void> {
  try {
    await updateFinancialIntelligencePipeline();
  } catch (e) {
    logger.warn({ operation, recommendationId, ...extra, err: e }, "intelligence_pipeline_refresh_failed");
  }
}

// Matches Python's RECOMMENDATIONS_CACHE_KEY = "global" — recommendations
// are global, not portfolio-scoped.
const RECOMMENDATIONS_CACHE_KEY = "global";

export interface SerializedRecommendation {
  id: string;
  asset_id: string;
  symbol: string | null;
  recommendation_state: string;
  confidence_score: number;
  status: string;
  version: string;
  created_at: string | null;
  updated_at: string | null;
  explanation: {
    rules_matched: unknown;
    reasoning: string;
    confidence_factors: unknown;
  } | null;
  outcome: {
    status: string;
    action_taken_at: string | null;
    dismiss_reason: string | null;
    ledger_transaction_id: string | null;
    predicted_impact: number | null;
    realized_impact: number | null;
  } | null;
}

/** Port of serialize_recommendation. */
export async function serializeRecommendation(rec: Recommendation): Promise<SerializedRecommendation> {
  const expl = await prisma.recommendation_explanations.findUnique({ where: { recommendation_id: rec.id } });
  const out = await prisma.recommendation_outcomes.findUnique({ where: { recommendation_id: rec.id } });
  const quote = await prisma.latestQuote.findFirst({ where: { assetId: rec.asset_id } });

  return {
    id: rec.id,
    asset_id: rec.asset_id,
    symbol: quote?.symbol ?? null,
    recommendation_state: rec.recommendation_state,
    confidence_score: Number(rec.confidence_score),
    status: rec.status,
    version: rec.version,
    created_at: rec.created_at?.toISOString() ?? null,
    updated_at: rec.updated_at?.toISOString() ?? null,
    explanation: expl
      ? { rules_matched: expl.rules_matched, reasoning: expl.reasoning, confidence_factors: expl.confidence_factors }
      : null,
    outcome: out
      ? {
          status: out.status,
          action_taken_at: out.action_taken_at?.toISOString() ?? null,
          dismiss_reason: out.dismiss_reason,
          ledger_transaction_id: out.ledger_transaction_id,
          predicted_impact: out.predicted_impact !== null ? Number(out.predicted_impact) : null,
          realized_impact: out.realized_impact !== null ? Number(out.realized_impact) : null,
        }
      : null,
  };
}

interface Features {
  momentum_score: unknown;
  volatility_score: unknown;
  sentiment_score: unknown;
}
interface Scores {
  quality_score: unknown;
  valuation_score: unknown;
}

/** Port of _score_and_materialize. Returns null without writing anything if
 * a required factor hasn't been computed yet — never fabricates a neutral
 * substitute for a missing value.
 *
 * Crypto has no fundamentals analog (FUNDAMENTALS_SCORING_SCOPE.md §6) so
 * quality_score/valuation_score are permanently null for it — the gate and
 * rule engine branch on asset_class so crypto materializes off
 * momentum/volatility/sentiment alone instead of staying dark forever. */
export async function scoreAndMaterialize(assetId: string, features: Features, scores: Scores): Promise<Recommendation | null> {
  const asset = await prisma.asset.findUnique({ where: { id: assetId } });
  const isCrypto = asset !== null && asset.assetClass === "crypto";

  const requiredFactors = isCrypto
    ? [features.momentum_score, features.volatility_score, features.sentiment_score]
    : [features.momentum_score, features.volatility_score, features.sentiment_score, scores.quality_score, scores.valuation_score];

  if (requiredFactors.some((f) => f === null || f === undefined)) return null;

  const momentum = Number(features.momentum_score);
  const volatility = Number(features.volatility_score);
  const sentiment = Number(features.sentiment_score);

  let recState = "HOLD";
  let reasoning = "Asset parameters remain within stable bounds. Recommending holding current position.";
  let rulesMatched: Prisma.InputJsonValue = {};
  let confidenceFactors: Prisma.InputJsonValue = {};
  let confidenceScore = 0.5;

  if (isCrypto) {
    if (momentum >= 0.5 && sentiment >= 0.5) {
      recState = "BUY";
      reasoning = "Asset displays positive momentum combined with constructive market sentiment (crypto: no valuation signal available).";
      rulesMatched = { momentum_and_sentiment: true };
      confidenceFactors = { momentum: 0.5, sentiment: 0.5 };
      confidenceScore = 0.5 * momentum + 0.5 * sentiment;
    } else if (sentiment < 0.3 && momentum < 0.4) {
      recState = "AVOID";
      reasoning = "Asset displays weak market sentiment and negative momentum, prompting caution.";
      rulesMatched = { weak_sentiment_and_momentum: true };
      confidenceFactors = { sentiment: 0.5, momentum: 0.5 };
      confidenceScore = 0.5 * (1.0 - sentiment) + 0.5 * (1.0 - momentum);
    } else if (volatility >= 0.6 && momentum < 0.4) {
      recState = "REDUCE";
      reasoning = "Asset displays weakening momentum with elevated volatility, recommending reducing exposure (crypto: no valuation signal available).";
      rulesMatched = { weakening_momentum_and_volatility: true };
      confidenceFactors = { volatility: 0.5, momentum: 0.5 };
      confidenceScore = 0.5 * volatility + 0.5 * (1.0 - momentum);
    } else {
      recState = "HOLD";
      reasoning = "Asset parameters remain within stable bounds. Recommending holding current position.";
      rulesMatched = { stable_parameters: true };
      confidenceFactors = { sentiment: 0.5, volatility: 0.5 };
      confidenceScore = 0.5 * sentiment + 0.5 * (1.0 - volatility);
    }
  } else {
    const quality = Number(scores.quality_score);
    const valuation = Number(scores.valuation_score);

    if (valuation >= 0.7 && momentum >= 0.5 && sentiment >= 0.5) {
      recState = "BUY";
      reasoning = "Asset displays strong underpricing combined with positive momentum and constructive market sentiment.";
      rulesMatched = { underpricing_and_momentum: true };
      confidenceFactors = { valuation: 0.4, momentum: 0.3, sentiment: 0.3 };
      confidenceScore = 0.4 * valuation + 0.3 * momentum + 0.3 * sentiment;
    } else if (sentiment < 0.3 && momentum < 0.4) {
      recState = "AVOID";
      reasoning = "Asset displays weak market sentiment and negative momentum, prompting caution.";
      rulesMatched = { weak_sentiment_and_momentum: true };
      confidenceFactors = { sentiment: 0.5, momentum: 0.5 };
      confidenceScore = 0.5 * (1.0 - sentiment) + 0.5 * (1.0 - momentum);
    } else if (valuation < 0.4 && volatility >= 0.6) {
      recState = "REDUCE";
      reasoning = "Asset is potentially overvalued with elevated volatility, recommending reducing exposure.";
      rulesMatched = { overvaluation_and_volatility: true };
      confidenceFactors = { valuation: 0.5, volatility: 0.5 };
      confidenceScore = 0.5 * (1.0 - valuation) + 0.5 * volatility;
    } else {
      recState = "HOLD";
      reasoning = "Asset parameters remain within stable bounds. Recommending holding current position.";
      rulesMatched = { stable_parameters: true };
      confidenceFactors = { quality: 0.5, volatility: 0.5 };
      confidenceScore = 0.5 * quality + 0.5 * (1.0 - volatility);
    }
  }

  const existingRec = await prisma.recommendations.findFirst({ where: { asset_id: assetId, version: "v2.0.0" } });
  const recId = existingRec?.id ?? uuidv4();
  const now = new Date();

  const rec = await prisma.recommendations.upsert({
    where: { id: recId },
    create: {
      id: recId,
      asset_id: assetId,
      recommendation_state: recState,
      confidence_score: confidenceScore,
      status: "active",
      version: "v2.0.0",
      created_at: now,
      updated_at: now,
    },
    update: {
      recommendation_state: recState,
      confidence_score: confidenceScore,
      updated_at: now,
    },
  });

  await prisma.recommendation_explanations.upsert({
    where: { recommendation_id: recId },
    create: {
      recommendation_id: recId,
      rules_matched: rulesMatched,
      reasoning,
      confidence_factors: confidenceFactors,
    },
    update: {
      rules_matched: rulesMatched,
      reasoning,
      confidence_factors: confidenceFactors,
    },
  });

  const existingOutcome = await prisma.recommendation_outcomes.findUnique({ where: { recommendation_id: recId } });
  if (!existingOutcome) {
    await prisma.recommendation_outcomes.create({
      data: { recommendation_id: recId, status: "active", action_taken_at: now },
    });
  }

  return rec;
}

/** Port of RecommendationRepository._held_asset_ids: distinct non-null
 * asset_ids held across every portfolio (this app is single-user/no-
 * multi-tenancy, so Recommendation stays asset-scoped rather than gaining a
 * portfolio_id column). Exported so other callers of Python's
 * RecommendationRepository.get_all() (which filters to held assets, not
 * every Recommendation row) can apply the same filter — see
 * lib/ai/intelligence.ts's updateFinancialIntelligencePipeline and
 * routes/ai/recommendations.ts's GET /recommendations list. */
export async function heldAssetIds(): Promise<string[]> {
  const rows = await prisma.position.findMany({
    where: { assetId: { not: null } },
    select: { assetId: true },
    distinct: ["assetId"],
  });
  return rows.map((r) => r.assetId!).filter((id): id is string => id !== null);
}

/** Port of generate_recommendations — the manual batch trigger. Iterates
 * AssetSnapshot rows for assets actually held in a portfolio position,
 * skipping any without both AssetFeatures and an AssetScore row yet. */
export async function generateRecommendations(): Promise<SerializedRecommendation[]> {
  const assetIds = await heldAssetIds();
  const snapshots = assetIds.length > 0 ? await prisma.assetSnapshot.findMany({ where: { assetId: { in: assetIds } } }) : [];
  const recsCreated: Recommendation[] = [];

  for (const snap of snapshots) {
    const assetId = snap.assetId;
    const features = await prisma.asset_features.findUnique({ where: { asset_id: assetId } });
    const scores = await prisma.assetScore.findFirst({
      where: { assetId },
      orderBy: { generatedAt: "desc" },
    });

    if (!features || !scores) continue;

    const rec = await scoreAndMaterialize(assetId, features, {
      quality_score: scores.qualityScore,
      valuation_score: scores.valuationScore,
    });
    if (rec !== null) recsCreated.push(rec);
  }

  return Promise.all(recsCreated.map((r) => serializeRecommendation(r)));
}

/** Port of apply_recommendation. Creates a real Transaction against the
 * target portfolio (or the first portfolio found, if none specified),
 * marks the recommendation "applied", and records a predicted_impact
 * (+/-0.05, or 0.0 for HOLD — a fixed heuristic, not a model output; see
 * Python's identical hardcoded values). */
export async function applyRecommendation(
  recommendationId: string,
  portfolioId: string | null,
  actorId: string | null,
): Promise<SerializedRecommendation> {
  let resolvedPortfolioId = "";
  const updatedRec = await prisma.$transaction(async (tx) => {
    const rec = await tx.recommendations.findUnique({ where: { id: recommendationId } });
    if (!rec) throw new NotFoundError("Recommendation not found");
    if (rec.status !== "active") throw new ValidationError(`Recommendation is already ${rec.status}`);

    let targetPortfolioId = portfolioId;
    if (!targetPortfolioId) {
      const portfolio = await tx.portfolio.findFirst();
      if (!portfolio) throw new ValidationError("No portfolios found to apply recommendation");
      targetPortfolioId = portfolio.id;
    } else {
      const portfolio = await tx.portfolio.findUnique({ where: { id: targetPortfolioId } });
      if (!portfolio) throw new ValidationError("Invalid portfolio");
    }

    const quote = await tx.latestQuote.findFirst({ where: { assetId: rec.asset_id } });
    const symbol = quote?.symbol ?? "UNKNOWN";
    const price = quote ? Number(quote.price) : 0.0;

    const isReduceOrAvoid = rec.recommendation_state === "REDUCE" || rec.recommendation_state === "AVOID";
    const transactionType = rec.recommendation_state === "BUY" ? "BUY" : isReduceOrAvoid ? "SELL" : "HOLD";
    const quantity = rec.recommendation_state === "BUY" || isReduceOrAvoid ? 1.0 : 0.0;
    const now = new Date();
    const txnId = uuidv4();

    await tx.transaction.create({
      data: {
        id: txnId,
        portfolioId: targetPortfolioId,
        symbol,
        assetId: rec.asset_id,
        transactionType,
        quantity,
        price,
        transactionDate: now,
        fees: 0.0,
        taxes: 0.0,
        notes: `Applied recommendation ${rec.id} (${rec.recommendation_state})`,
        broker: "aureon",
        kind: "trade",
        recommendationId: rec.id,
        createdAt: now,
        updatedAt: now,
      },
    });

    const updated = await tx.recommendations.update({
      where: { id: recommendationId },
      data: { status: "applied", updated_at: now },
    });

    const predictedImpact = rec.recommendation_state === "BUY" ? 0.05 : isReduceOrAvoid ? -0.05 : 0.0;

    await tx.recommendation_outcomes.upsert({
      where: { recommendation_id: recommendationId },
      create: {
        recommendation_id: recommendationId,
        status: "applied",
        action_taken_at: now,
        ledger_transaction_id: txnId,
        predicted_impact: predictedImpact,
      },
      update: {
        status: "applied",
        action_taken_at: now,
        ledger_transaction_id: txnId,
        predicted_impact: predictedImpact,
      },
    });

    await logAuditAction(tx, "recommendation_apply", "recommendation", actorId, recommendationId, {
      recommendation_state: rec.recommendation_state,
      portfolio_id: targetPortfolioId,
    });

    resolvedPortfolioId = targetPortfolioId;
    return updated;
  });

  await invalidateOrgRecommendations(RECOMMENDATIONS_CACHE_KEY);
  await refreshIntelligencePipeline("apply_recommendation", recommendationId, { portfolio_id: resolvedPortfolioId });
  return serializeRecommendation(updatedRec);
}

/** Port of dismiss_recommendation. */
export async function dismissRecommendation(
  recommendationId: string,
  reason: string | null,
  actorId: string | null,
): Promise<SerializedRecommendation> {
  const updatedRec = await prisma.$transaction(async (tx) => {
    const rec = await tx.recommendations.findUnique({ where: { id: recommendationId } });
    if (!rec) throw new NotFoundError("Recommendation not found");
    if (rec.status !== "active") throw new ValidationError(`Recommendation is already ${rec.status}`);

    const now = new Date();
    const updated = await tx.recommendations.update({
      where: { id: recommendationId },
      data: { status: "dismissed", updated_at: now },
    });

    await tx.recommendation_outcomes.upsert({
      where: { recommendation_id: recommendationId },
      create: {
        recommendation_id: recommendationId,
        status: "dismissed",
        action_taken_at: now,
        dismiss_reason: reason,
      },
      update: {
        status: "dismissed",
        action_taken_at: now,
        dismiss_reason: reason,
      },
    });

    await logAuditAction(tx, "recommendation_dismiss", "recommendation", actorId, recommendationId, { reason });

    return updated;
  });

  await invalidateOrgRecommendations(RECOMMENDATIONS_CACHE_KEY);
  await refreshIntelligencePipeline("dismiss_recommendation", recommendationId);
  return serializeRecommendation(updatedRec);
}

/** Port of undo_recommendation. Nulls the outcome's ledger_transaction_id
 * before deleting the Transaction row (order matters: recommendation_outcomes
 * has a real FK to transactions, so the reference must be cleared first —
 * Python's SQLAlchemy unit-of-work reorders these automatically within one
 * session.commit(), Prisma does not). */
export async function undoRecommendation(recommendationId: string, actorId: string | null): Promise<SerializedRecommendation> {
  const updatedRec = await prisma.$transaction(async (tx) => {
    const rec = await tx.recommendations.findUnique({ where: { id: recommendationId } });
    if (!rec) throw new NotFoundError("Recommendation not found");
    if (rec.status === "active") throw new ValidationError("Recommendation is already active");

    const out = await tx.recommendation_outcomes.findUnique({ where: { recommendation_id: recommendationId } });
    const ledgerTransactionId = out?.ledger_transaction_id ?? null;
    const now = new Date();

    const updated = await tx.recommendations.update({
      where: { id: recommendationId },
      data: { status: "active", updated_at: now },
    });

    if (out) {
      await tx.recommendation_outcomes.update({
        where: { recommendation_id: recommendationId },
        data: { status: "active", action_taken_at: now, ledger_transaction_id: null, dismiss_reason: null },
      });
    }

    if (ledgerTransactionId) {
      const txn = await tx.transaction.findUnique({ where: { id: ledgerTransactionId } });
      if (txn) await tx.transaction.delete({ where: { id: txn.id } });
    }

    // Matches Python exactly: rec.status is read here *after* being mutated
    // to "active" above (recommendation.py's undo_recommendation logs
    // rec.status post-mutation too), so this always logs "active" rather
    // than the actual prior status — a pre-existing quirk in the Python
    // source, ported as-is rather than silently fixed.
    await logAuditAction(tx, "recommendation_undo", "recommendation", actorId, recommendationId, {
      previous_status: updated.status,
    });

    return updated;
  });

  await invalidateOrgRecommendations(RECOMMENDATIONS_CACHE_KEY);
  await refreshIntelligencePipeline("undo_recommendation", recommendationId);
  return serializeRecommendation(updatedRec);
}
